test/indexer_test.c

```diff
@@ -0,0 +1,113 @@
+#include "test.h"
+
+#include "index.c"
+
+enum {
+    CAP = 10
+};
+
+bool test_indexer_build(void)
+{
+    struct indexer *indexer = indexer_alloc(CAP);
+    assert(indexer);
+    assert(indexer->len == 0);
+    assert(indexer->cap == CAP);
+
+    rill_key_t data[CAP] = {0, 2, 4, 6, 8, 10, 12, 14, 16, 18};
+    for (size_t i = 0; i < CAP; i++)
+        indexer_put(indexer, data[i], i);
+    assert(indexer->len == indexer->cap);
+
+    const size_t size = indexer_cap(CAP);
+    struct index *index = calloc(1, size);
+    assert(index);
+
+    size_t n_written = indexer_write(indexer, index);
+    assert(n_written == size);
+    assert(index->slope == (data[CAP - 1] - data[0]) / CAP);
+
+    indexer_free(indexer);
+
+    for (size_t i = 0; i < index->len; i++)
+        assert(index_get(index, i) == data[i]);
+    assert(index_get(index, index->len) == 0);
+
+    free(index);
+    
+    return true;
+}
+
+static struct index *make_index(rill_key_t *data, size_t n) 
+{
+    struct indexer *indexer = indexer_alloc(n);
+    for (size_t i = 0; i < n; i++)
+        indexer_put(indexer, data[i], i);
+
+    struct index *index = calloc(1, indexer_cap(n));
+    indexer_write(indexer, index);
+    indexer_free(indexer);
+
+    return index;
+}
+
+static void check_find_all(struct index *index, rill_key_t *data, size_t n)
+{
+    size_t key_idx;
+    uint64_t val;
+
+    for (size_t i = 0; i < n; i++) {
+        assert(index_find(index, data[i], &key_idx, &val));
+        assert(key_idx == i);
+        assert(val == i);
+    }
+}
+
+bool test_indexer_lookup(void)
+{
+    struct index *index;
+
+    rill_key_t data1[CAP] = {0, 3, 6, 9, 12, 15, 18, 21, 24, 27};
+    index = make_index(data1, CAP);
```

Comment:
You can avoid having to create a temporary array and variable by wrapping the make_index call ina macro like so: https://github.com/RAttab/rill/blob/master/test/coder_test.c#L61

I don't recommend doing that for regular code, but for tests, it significantly reduces the boiler plate.